src/Collection.ts

```diff
@@ -120,25 +155,42 @@ export default abstract class Collection<T extends Model> extends Base {
   }
 
   /**
-   * Get resources matching criteria
+   * Get resources matching criteria.
+   *
+   * If passing an object of key:value conditions, it will
+   * use the indexes to efficiently retrieve the data.
    */
   filter (query: { [key: string]: any } | ((T) => boolean)): Array<T> {
-    return this.models.filter(model => {
-      return typeof query === 'function'
-        ? query(model)
-        : isMatch(model.toJS(), query)
-    })
+    if (typeof query === 'function') {
+      return this.models.filter(model => query(model))
+    } else {
+      // Sort the query to hit the indexes first
+      const optimizedQuery = entries(query).sort((A, B) =>
+        Number(this.index.has(B[0])) - Number(this.index.has(A[0]))
+      )
```

Comment:
What if instead of doing the sorting in every filter delegate the sorting to developer? I think it's not hard to remember to add first indexed attributes. This way we can avoid this sort

Maybe it's not a relevant performance improvement but you could check it how sorted vs unsorted performs :)

Reply:
Sorting object keys isn't as straightforward as it seems: https://stackoverflow.com/questions/5467129/sort-javascript-object-by-key

This is why I'm first converting the hash into an array of entries.

Reply:
also, this gives us more performance out of the box without having to manually tune anything. Defaults are important.